apollo-gradle-plugin-incubating/src/main/kotlin/com/apollographql/apollo/gradle/internal/ApolloGenerateSourcesTask.kt

```diff
@@ -18,6 +19,10 @@ abstract class ApolloGenerateSourcesTask : DefaultTask() {
   @get:Optional
   abstract val customTypeMapping: MapProperty<String, String>
 
+  @get:Input
+  @get:Optional
+  abstract val customIdGenerator: Property<String>
```

Comment:
Thinking out loud: what about just making this? 

```kotlin
  @get:Optional
  abstract val customIdGenerator: CustomIDGenerator
```

Then we would rely on the fact that changing the `CustomIDGenerator` would invalidate the whole build as it would change the build scripts somehow. I'm not sure about that but it should, right ? In all cases, it's not too different than the current String property + reflection and provides compile time type safety.

Reply:
If we use Optional, Internal, how can we pass the IdGenerator from the build.gradle file? 

I tried this change and the tests fail with `Could not generate a decorated class for class com.apollographql.apollo.gradle.internal.ApolloGenerateSourcesTask` .

Reply:
Oopps, sorry, the property shouldn't be abstract in that case, I edited my comment above

Reply:
I don't understand how we can pass `customIdGenerator` to Apollo's task `:generateApolloSources` if this is not an `Input`.

Reply:
`Input` is used for up-to-date checks but ultimately, `ApolloGenerateSourcesTask` is a regular java class so you can set fields like you would do in regular Kotlin:

```kotlin
// ApolloPlugin.kt
     project.tasks.register(taskName, ApolloGenerateSourcesTask::class.java) {
        it.customIdGenerator = apolloExtension.customIdGenerator
        ...
```

```build.gradle.kts
apollo {
   customIdGenerator = {transformedQuery ->
       myHashFunction(transformedQuery)
   }
}
```

Reply:
Thank you. I updated both the plugins to use this pattern and pushed the changes.

Reply:
The thing is I'm not sure how this plays with up-to-date checks, cacheability, etc... I guess we'll have to add tests for that.

Reply:
Can you help me write these tests?

Should I overwrite the existing integration tests' `testProject/.../build.gradle` to use a custom id generator or should I create a new project for testing this?

Reply:
I think this would break cacheability. And it is not easy to write tests for build cache behavior.

Reply:
Turns out gradle 6 doesn't like Optional + Internal. The CI currently fails with:

```
Property 'customIdGenerator' getter 'getCustomIdGenerator()' annotated with @Internal should not be also annotated with @Optional. This behaviour has been deprecated and is scheduled to be removed in Gradle 7.0.
```
So I guess it's just`Internal` then ?